src/Lecture2.hs

```diff
@@ -154,13 +167,161 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
+
+newtype Health = Health Int
+
+mkHealth :: Int -> Maybe Health
+mkHealth value
+  | value < 0 = Nothing
+  | otherwise = Just $ Health value
+
+unHealth :: Health -> Int
+unHealth (Health health) = health
+
+newtype AttackPower = AttackPower Int
+
+mkAttackPower :: Int -> Maybe AttackPower
+mkAttackPower value
+  | value < 0 = Nothing
+  | otherwise = Just $ AttackPower value
+
+unAttackPower :: AttackPower -> Int
+unAttackPower (AttackPower power) = power
+
+newtype Endurance = Endurance Int
+
+mkEndurance :: Int -> Maybe Endurance
+mkEndurance value
+  | value < 0 = Nothing
+  | otherwise = Just $ Endurance value
+
+unEndurance :: Endurance -> Int
+unEndurance (Endurance endurance) = endurance
+
+data Treasure = Treasure
+
+newtype Gold = Gold Int
+
+mkGold :: Int -> Maybe Gold
+mkGold value
+  | value < 0 = Nothing
+  | otherwise = Just $ Gold value
+
+unGold :: Gold -> Int
+unGold (Gold gold) = gold
+
+newtype Experience = Experience Int
+
+mkExperience :: Int -> Maybe Experience
+mkExperience value
+  | value < 0 = Nothing
+  | otherwise = Just $ Experience value
+
+unExperience :: Experience -> Int
+unExperience (Experience experience) = experience
+
 data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
+    { knightHealth    :: Health
+    , knightAttack    :: AttackPower
+    , knightEndurance :: Endurance
     }
 
-dragonFight = error "TODO"
+data KnightStats = KnightStats { ksH :: Int
+                               , ksP :: Int
+                               , ksE :: Int }
+
+mkKnight :: KnightStats -> Maybe Knight
+mkKnight stats = case (mkHealth $ ksH stats, mkAttackPower $ ksP stats, mkEndurance $ ksE stats) of
+  (Just h, Just p, Just e) -> Just $ Knight { knightHealth    = h
+                                            , knightAttack    = p
+                                            , knightEndurance = e }
+  _                        -> Nothing
+
+data DragonInfo = DragonInfo
+  { dragonHealth    :: Health
+  , dragonFirePower :: AttackPower
+  , dragonGold      :: Gold}
+
+data DragonInfoStats = DragonInfoStats { disH :: Int
+                                       , disP :: Int
+                                       , disG :: Int }
+
+mkDragonInfo :: DragonInfoStats -> Maybe DragonInfo
+mkDragonInfo stats = case (mkHealth $ disH stats, mkAttackPower $ disP stats, mkGold $ disG stats) of
+  (Just h, Just p, Just g) -> Just $ DragonInfo { dragonHealth     = h
+                                                , dragonFirePower = p
+                                                , dragonGold      = g}
+  _                        -> Nothing
+
+unDragonInfo :: Dragon -> DragonInfo
+unDragonInfo (Red _ info)   = info
+unDragonInfo (Black _ info) = info
+unDragonInfo (Green info)   = info
+
+data Dragon
+  = Red (Maybe Treasure) DragonInfo
+  | Black (Maybe Treasure) DragonInfo
+  | Green DragonInfo
```

Comment:
That's a correct implementation that makes invalid states unrepresentable 👍🏻 

If you want to introduce polymorphic types for treasure, you can easily change your existing data type:

```suggestion
data Chest a = Chest
    { chestGold :: Gold
    , chestTreasure :: a
    }

data Dragon a
  = Red (Chest a) DragonInfo
  | Black (Chest a) DragonInfo
  | Green Gold DragonInfo
```

I see that you've put `Gold` inside `DragonInfo`. That's also a valid solution 🙂 
Under these requirements, there's no clear winner of one solution over another.